### ISSUE TITLE
Mute monitor logs from AccumulatingProgressMonitor.beginTask

### ISSUE THREAD

**Issue by iloveeclipse** [eclipse-platform/eclipse.platform.ui#1158]
Follow up on https://github.com/eclipse-platform/eclipse.platform.ui/pull/1115.

While running latest I Build with our application tests I observe a lot of cases where clients can't do anything meaningful to prevent error log polluted with "beginTask should only be called once per instance" warnings from AccumulatingProgressMonitor.

In the stack below framework started beginTask in the progress dialog while waiting on a job and in the job client code simply used given job monitor to call SubMonitor.convert(monitor). This is fully valid use case, but there is no way to avoid the warning logged.

```
!ENTRY org.eclipse.jface 2 0 2023-09-20 20:32:21.203
!MESSAGE beginTask should only be called once per instance. At least call done() before further invocations
!STACK 0
java.lang.IllegalStateException: beginTask should only be called once per instance. At least call done() before further invocations
	at org.eclipse.jface.operation.AccumulatingProgressMonitor.beginTask(AccumulatingProgressMonitor.java:148)
	at org.eclipse.ui.internal.progress.ProgressManager$JobMonitor.lambda$0(ProgressManager.java:265)
	at java.base/java.lang.Iterable.forEach(Iterable.java:75)
	at java.base/java.util.Collections$UnmodifiableCollection.forEach(Collections.java:1092)
	at org.eclipse.ui.internal.progress.ProgressManager$JobMonitor.beginTask(ProgressManager.java:265)
	at org.eclipse.core.runtime.SubMonitor.convert(SubMonitor.java:483)
	at org.eclipse.core.runtime.SubMonitor.convert(SubMonitor.java:449)
	at some.client.method()
	at org.eclipse.core.internal.resources.InternalWorkspaceJob.run(InternalWorkspaceJob.java:43)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
Caused by: java.lang.IllegalStateException: beginTask(, 1000) was called here previously
	at org.eclipse.jface.operation.AccumulatingProgressMonitor.beginTask(AccumulatingProgressMonitor.java:155)
	at org.eclipse.ui.internal.progress.ProgressManager$JobMonitor.addProgressListener(ProgressManager.java:244)
	at org.eclipse.ui.internal.progress.ProgressMonitorFocusJobDialog$3.runInUIThread(ProgressMonitorFocusJobDialog.java:213)
	at org.eclipse.ui.progress.UIJob.lambda$0(UIJob.java:148)
	at org.eclipse.swt.widgets.RunnableLock.run(RunnableLock.java:40)
	at org.eclipse.swt.widgets.Synchronizer.runAsyncMessages(Synchronizer.java:132)
	at org.eclipse.swt.widgets.Display.runAsyncMessages(Display.java:5040)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:4520)
```

Another case is worse. Again, client code starts to work, calls beginTask() at very beginning of a long operation and later on passes given monitor to IResource.refreshLocal() which calls SubMonitor.convert(monitor) and that calls beginTask() that triggers a warning again. Once again, it is absolutely valid to pass any given monitor to IResource.refreshLocal() and not doing this would break the cancellation of the possibly long running operation.

```
java.lang.IllegalStateException: beginTask should only be called once per instance. At least call done() before further invocations
	at org.eclipse.jface.operation.AccumulatingProgressMonitor.beginTask(AccumulatingProgressMonitor.java:148)
	at org.eclipse.ui.internal.progress.ProgressManager$JobMonitor.lambda$0(ProgressManager.java:265)
	at java.base/java.lang.Iterable.forEach(Iterable.java:75)
	at java.base/java.util.Collections$UnmodifiableCollection.forEach(Collections.java:1092)
	at org.eclipse.ui.internal.progress.ProgressManager$JobMonitor.beginTask(ProgressManager.java:265)
	at org.eclipse.core.runtime.SubMonitor.convert(SubMonitor.java:483)
	at org.eclipse.core.runtime.SubMonitor.convert(SubMonitor.java:449)
	at org.eclipse.core.internal.resources.Resource.refreshLocal(Resource.java:1596)
	at some.client.code.calling.refreshLocal()
	at org.eclipse.core.internal.resources.InternalWorkspaceJob.run(InternalWorkspaceJob.java:43)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
Caused by: java.lang.IllegalStateException: beginTask(, 1000) was called here previously
	at org.eclipse.jface.operation.AccumulatingProgressMonitor.beginTask(AccumulatingProgressMonitor.java:155)
	at org.eclipse.ui.internal.progress.ProgressManager$JobMonitor.lambda$0(ProgressManager.java:265)
	at java.base/java.lang.Iterable.forEach(Iterable.java:75)
	at java.base/java.util.Collections$UnmodifiableCollection.forEach(Collections.java:1092)
	at org.eclipse.ui.internal.progress.ProgressManager$JobMonitor.beginTask(ProgressManager.java:265)
	at org.eclipse.core.runtime.SubMonitor.convert(SubMonitor.java:483)
	at org.eclipse.core.runtime.SubMonitor.convert(SubMonitor.java:449)
	at some.client.code.called.from.WorkspaceJob.runInWorkspace()
	at org.eclipse.core.internal.resources.InternalWorkspaceJob.run(InternalWorkspaceJob.java:43)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
	... 2 more
```

So basically clients are unable now to pass monitors to any Resource related API without getting a warning logged.
I haven't analyzed all the warnings yet (there are literally hundreds of them), but that above alone is enough.

So I would propose to mute the logging from AccumulatingProgressMonitor by default and enable it on demand with additional VM option.

I will provide a PR.

**Comment by jukzi** [comment]
> In the stack below framework started beginTask in the progress dialog while waiting on a job and in the job client code simply used given job monitor to call SubMonitor.convert(monitor). This is fully valid use case,

That is explicitly forbidden by javadoc
`<li>Will call {@link #beginTask} on the argument 0 or 1 times, at its option.</li>` see IProgressMonitor:53 ff

If you have and example - preferably junit - i could investigate if it could be fixed, until now i have seen only very obvious wrong usages.

Nevertheless we could disable (or completely remove) the logging later, but by now i would prefer to fix the wrong calls.

**Comment by iloveeclipse** [comment]
> That is explicitly forbidden by javadoc

Jörg, the resource API receives a monitor and unconditionally calls beginTask(). The client that passes this monitor received it from framework with started task already. The violation is here "built in".

> but by now i would prefer to fix the wrong calls.

There is nothing to fix!

**Comment by iloveeclipse** [comment]
In short: clients using resource API now can't start any task and pass this monitor to resources API without receiving warning logged. This is insane.

**Comment by jukzi** [comment]
please give an example. if the error is in the framework it might be possible to fix it there.

**Comment by iloveeclipse** [comment]
See stack traces above. You can't fix that without breaking something else.

**Comment by jukzi** [comment]
please give an example how to create that stack traces.

**Comment by iloveeclipse** [comment]
But independently if the problem can or can't be fixed, if you want track monitor use violations, you still can enable the logging with https://github.com/eclipse-platform/eclipse.platform.ui/pull/1159.

**Comment by iloveeclipse** [comment]
Please apply attached patch to `org.eclipse.ui.examples.job` project, start Eclipse from debugger, open "Job Factory" view, toggle "user job" and click on "Create jobs" button. It will flood the error log with warnings.

What the patch does is fully legal and nothing is wrong in the client code.

![image](https://github.com/eclipse-platform/eclipse.platform.ui/assets/964108/91843c3d-608d-4176-97e1-0846b482cafc)

```
diff --git a/examples/org.eclipse.ui.examples.job/src/org/eclipse/ui/examples/jobs/TestJob.java b/examples/org.eclipse.ui.examples.job/src/org/eclipse/ui/examples/jobs/TestJob.java
index 7e05e8c..63cdeef 100644
--- a/examples/org.eclipse.ui.examples.job/src/org/eclipse/ui/examples/jobs/TestJob.java
+++ b/examples/org.eclipse.ui.examples.job/src/org/eclipse/ui/examples/jobs/TestJob.java
@@ -17,2 +17,3 @@
 import org.eclipse.core.resources.ResourcesPlugin;
+import org.eclipse.core.runtime.CoreException;
 import org.eclipse.core.runtime.IProgressMonitor;
@@ -101,2 +102,7 @@
 			for (int i = 0; i < ticks; i++) {
+				try {
+					ResourcesPlugin.getWorkspace().getRoot().refreshLocal(0, monitor);
+				} catch (CoreException e) {
+					e.printStackTrace();
+				}
 				if (monitor.isCanceled()) {
diff --git a/examples/org.eclipse.ui.examples.job/src/org/eclipse/ui/examples/jobs/views/JobsView.java b/examples/org.eclipse.ui.examples.job/src/org/eclipse/ui/examples/jobs/views/JobsView.java
index 29ad113..8c5532e 100644
--- a/examples/org.eclipse.ui.examples.job/src/org/eclipse/ui/examples/jobs/views/JobsView.java
+++ b/examples/org.eclipse.ui.examples.job/src/org/eclipse/ui/examples/jobs/views/JobsView.java
@@ -27,2 +27,3 @@
 import org.eclipse.core.runtime.jobs.Job;
+import org.eclipse.core.runtime.preferences.InstanceScope;
 import org.eclipse.jface.action.Action;
@@ -45,2 +46,3 @@
 import org.eclipse.ui.examples.jobs.UITestJob;
+import org.eclipse.ui.internal.IPreferenceConstants;
 import org.eclipse.ui.part.ViewPart;
@@ -81,2 +83,4 @@
 	protected void createJobs() {
+		InstanceScope.INSTANCE.getNode("org.eclipse.ui.workbench").putBoolean(IPreferenceConstants.RUN_IN_BACKGROUND, //$NON-NLS-1$
+				false);
 		int jobCount = Integer.parseInt(quantityField.getText());

```

**Comment by iloveeclipse** [comment]
The patch above, extended to another valid case (I'm still digging through various logs reported by tests):

```
diff --git a/examples/org.eclipse.ui.examples.job/src/org/eclipse/ui/examples/jobs/TestJob.java b/examples/org.eclipse.ui.examples.job/src/org/eclipse/ui/examples/jobs/TestJob.java
index 7e05e8c..4e4233e 100644
--- a/examples/org.eclipse.ui.examples.job/src/org/eclipse/ui/examples/jobs/TestJob.java
+++ b/examples/org.eclipse.ui.examples.job/src/org/eclipse/ui/examples/jobs/TestJob.java
@@ -16,3 +16,6 @@
 
+import org.eclipse.core.resources.IWorkspace;
+import org.eclipse.core.resources.IWorkspaceRunnable;
 import org.eclipse.core.resources.ResourcesPlugin;
+import org.eclipse.core.runtime.CoreException;
 import org.eclipse.core.runtime.IProgressMonitor;
@@ -101,2 +104,21 @@
 			for (int i = 0; i < ticks; i++) {
+				// First case: just any resource operation
+				try {
+					ResourcesPlugin.getWorkspace().getRoot().refreshLocal(0, monitor);
+				} catch (CoreException e) {
+					e.printStackTrace();
+				}
+				// Second case: just call ResourcesPlugin.getWorkspace().run()
+				try {
+					ResourcesPlugin.getWorkspace().run(new IWorkspaceRunnable() {
+
+						@Override
+						public void run(IProgressMonitor innerMonitor) {
+							//
+						}
+					}, ResourcesPlugin.getWorkspace().getRoot(), IWorkspace.AVOID_UPDATE, monitor);
+				} catch (CoreException e) {
+					e.printStackTrace();
+				}
+
 				if (monitor.isCanceled()) {
diff --git a/examples/org.eclipse.ui.examples.job/src/org/eclipse/ui/examples/jobs/views/JobsView.java b/examples/org.eclipse.ui.examples.job/src/org/eclipse/ui/examples/jobs/views/JobsView.java
index 29ad113..8c5532e 100644
--- a/examples/org.eclipse.ui.examples.job/src/org/eclipse/ui/examples/jobs/views/JobsView.java
+++ b/examples/org.eclipse.ui.examples.job/src/org/eclipse/ui/examples/jobs/views/JobsView.java
@@ -27,2 +27,3 @@
 import org.eclipse.core.runtime.jobs.Job;
+import org.eclipse.core.runtime.preferences.InstanceScope;
 import org.eclipse.jface.action.Action;
@@ -45,2 +46,3 @@
 import org.eclipse.ui.examples.jobs.UITestJob;
+import org.eclipse.ui.internal.IPreferenceConstants;
 import org.eclipse.ui.part.ViewPart;
@@ -81,2 +83,4 @@
 	protected void createJobs() {
+		InstanceScope.INSTANCE.getNode("org.eclipse.ui.workbench").putBoolean(IPreferenceConstants.RUN_IN_BACKGROUND, //$NON-NLS-1$
+				false);
 		int jobCount = Integer.parseInt(quantityField.getText());

```

**Comment by iloveeclipse** [comment]
Found yet another case of valid code that now logs a warning.

For the easy of use, see this PR: https://github.com/eclipse-platform/eclipse.platform.ui/pull/1161

**Comment by iloveeclipse** [comment]
BTW, another problem I see is that using 3rd party bundles that are not "OK" for the new platform - if they can't be updated (for any reason, like incompatibility with other bundles or platform or other bugs or simply no new version available), the warning will be not fixable. See it now with egit that currently locked at specific version in our product. I see warnings that we've fixed on egit master but there is no release yet...

**Comment by jukzi** [comment]
The example you provided is not code in the IDE. So there is nothing to fix in the framework. You could change your example by using  submonitor:

```
diff --git a/examples/org.eclipse.ui.examples.job/src/org/eclipse/ui/examples/jobs/TestJob.java b/examples/org.eclipse.ui.examples.job/src/org/eclipse/ui/examples/jobs/TestJob.java
index ec43d8e..bfa9627 100644
--- a/examples/org.eclipse.ui.examples.job/src/org/eclipse/ui/examples/jobs/TestJob.java
+++ b/examples/org.eclipse.ui.examples.job/src/org/eclipse/ui/examples/jobs/TestJob.java
@@ -23,6 +23,7 @@
 import org.eclipse.core.runtime.MultiStatus;
 import org.eclipse.core.runtime.OperationCanceledException;
 import org.eclipse.core.runtime.Status;
+import org.eclipse.core.runtime.SubMonitor;
 import org.eclipse.core.runtime.jobs.Job;
 import org.eclipse.jface.resource.ResourceLocator;
 import org.eclipse.ui.progress.IProgressConstants;
@@ -94,18 +95,20 @@
 					new RuntimeException("This is the child exception"))); //$NON-NLS-1$
 			return result;
 		}
-		final long sleep = 10;
+		final long sleep = 1000;
 		int ticks = (int) (duration / sleep);
+		SubMonitor subMonitor;
 		if (this.unknown) {
-			monitor.beginTask(toString(), IProgressMonitor.UNKNOWN);
+			subMonitor = SubMonitor.convert(monitor, IProgressMonitor.UNKNOWN);
+			// monitor.beginTask(toString(), IProgressMonitor.UNKNOWN);
 		} else {
-			monitor.beginTask(toString(), ticks);
+			subMonitor = SubMonitor.convert(monitor, ticks * 4);
 		}
 		try {
 			for (int i = 0; i < ticks; i++) {
 				// First case: just any resource operation
 				try {
-					ResourcesPlugin.getWorkspace().getRoot().refreshLocal(0, monitor);
+					ResourcesPlugin.getWorkspace().getRoot().refreshLocal(0, subMonitor.split(1));
 				} catch (CoreException e) {
 					e.printStackTrace();
 				}
@@ -117,13 +120,13 @@
 						public void run(IProgressMonitor innerMonitor) {
 							//
 						}
-					}, ResourcesPlugin.getWorkspace().getRoot(), IWorkspace.AVOID_UPDATE, monitor);
+					}, ResourcesPlugin.getWorkspace().getRoot(), IWorkspace.AVOID_UPDATE, subMonitor.split(1));
 				} catch (CoreException e) {
 					e.printStackTrace();
 				}
 				// Third case: just call a join...
 				try {
-					Job.getJobManager().join(ResourcesPlugin.FAMILY_AUTO_BUILD, monitor);
+					Job.getJobManager().join(ResourcesPlugin.FAMILY_AUTO_BUILD, subMonitor.split(1));
 				} catch (OperationCanceledException | InterruptedException e) {
 					e.printStackTrace();
 				}
@@ -131,13 +134,13 @@
 				if (monitor.isCanceled()) {
 					return Status.CANCEL_STATUS;
 				}
-				monitor.subTask("Processing tick #" + i); //$NON-NLS-1$
+				monitor.setTaskName("Processing tick #" + i + "/" + ticks); //$NON-NLS-1$ //$NON-NLS-2$
 				try {
 					Thread.sleep(sleep);
 				} catch (InterruptedException e) {
 					return Status.CANCEL_STATUS;
 				}
-				monitor.worked(1);
+				subMonitor.worked(1);
 			}
 		} finally {
 			monitor.done();

```

**Comment by iloveeclipse** [comment]
> The example you provided is not code in the IDE. So there is nothing to fix in the framework. You could change your example by using submonitor:

1) You probably misunderstand something. I've provided examples of **valid** existing client code that doesn't work anymore without creating a warning. The code is client code, sure but that's purpose of Eclipse: a **platform** for various applications to build on it. The point that platform itself doesn't show such warnings doesn't mean it is not platform issue if logs are filled with warnings. This is NOT OK, it is a regression.

2) What you propose is that all the callers of `org.eclipse.core.runtime.IProgressMonitor.beginTask()` are supposed now to refactor the code just because the progress monitor they get could be `AccumulatingProgressMonitor`. If they don't doe that, they might see log flooded with warnings. This is again not OK.

3) We have ~900 references to `org.eclipse.core.runtime.IProgressMonitor.beginTask(String, int)` in my SDK workspace and ~800 in our product. It is not a viable solution to analyze & refactor all this code, especially there could be affected 3rd party bundles neither related to SDK nor to the application code.

4) The replacement code is not identical (the first task name set is missing on UI), but that is a minor issue & could be fixed if one would refactor all the client code.

**Comment by jukzi** [comment]
Not all beginTask() are forbidden, only repeated use of beginTask(). You have provided code with repeated beginTask(). That was always be disallowed by javadoc and could lead to a UI freeze. Now that produces a warning instead.

**Comment by iloveeclipse** [comment]
> You have provided code with repeated beginTask().

Not the user code repeated this, that was **platform code**!
Once again: it is fully valid to do on the client side in all 3 cases! Only because some platform code later down the stack calls beginTask() the problem appears.

**Comment by jukzi** [comment]
```
 * Methods that receive an {@link IProgressMonitor} ("callees") must either obey the following
 * conventions or include JavaDoc explaining how it differs from these rules.
 * The called method:
 * <ul>
 * <li>Will call {@link #beginTask} on the argument 0 or 1 times, at its option.</li>
...
 * The caller:
 * <ul>
 * <li>Will either pass in a fresh instance of {@link IProgressMonitor} that has
 *     not had {@link #beginTask} invoked on it yet, or will select an
 *     implementation of {@link IProgressMonitor} that explicitly permits multiple calls
 *     to {@link #beginTask}.</li>

```

**Comment by laeubi** [comment]
@jukzi I think an important part is

> implementation of {@link IProgressMonitor} **that explicitly permits multiple calls**  to {@link #beginTask}

with the recent changes I really wonder if/why one should impose that restriction or if we can't simply allow multiple `beginTask` calls for most of the implementations.

I think the main concern with `beginTask` called multiple times is that one passes the amount of work here, so the worst that can happen is that progress reporting is insufficient, but we can probably just call `setTaskName` and ignore the passed in amount of work if the monitor is already started.
